### PR TITLE
Permissionless TokenDistributor

### PR DESCRIPTION
**Pull request by ghostffcode**
- Upgrade openzeppelin contracts to v4
- Implement `safeERC20` for token transfers
- Use `call` for ETH transfer instead of `transfer`. See [https://solidity-by-example.org/sending-ether/](https://solidity-by-example.org/sending-ether/)

### CODE REVIEW

**Comment by varunv5000** [comment]
Is it necessary to have a "room" variable? Does this solve any pain points since we are already able to gather transaction data for a room.

Just wanted to know, it seems like it may help us retrieve room info directly from the blockchain, so seems like a good idea.

**Comment by ghostffcode** [comment]
> Is it necessary to have a "room" variable? Does this solve any pain points since we are already able to gather transaction data for a room.
> 
> 
> 
> Just wanted to know, it seems like it may help us retrieve room info directly from the blockchain, so seems like a good idea.

Yup, pretty much the reason behind it. Also indexing it makes for better record keeping

**Comment by varunv5000** [comment]
Tested the contract, seems everything is working fine, also, do we need to make the contract Ownable? Making it ownable seems to add extra functions that the contract doesn't seem to need.

**Review by varunv5000 — APPROVED** [review]

**Review by farque65 — APPROVED** [review]

File: packages/hardhat/deploy/00_deploy_tokendistributor_contract.js [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+  // transfer ownership to UI owner if needed
+  const ownershipTransfer = await tokenDistributorContract.transferOwnership(
+    developerAddress
+  );
+
+  await ownershipTransfer.wait(confirmationRequirement);
+
+  // Verify your contracts with Etherscan
+  // You don't want to verify on localhost
+  if (chainId !== localChainId) {
```
Comment (varunv5000) [inline packages/hardhat/deploy/00_deploy_tokendistributor_contract.js:57]: Deploying to etherscan doesn't work for me. The error I get is "ReferenceError: sleep is not defined".  Had to comment this out when deploying to rinkeby

File: packages/hardhat/deploy/00_deploy_tokendistributor_contract.js [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    // wait for etherscan to be ready to verify
+    await sleep(15000);
```
Comment (varunv5000) [inline packages/hardhat/deploy/00_deploy_tokendistributor_contract.js:59]: Deploying to etherscan doesn't work for me. The error I get is "ReferenceError: sleep is not defined". Had to comment this out when deploying to rinkeby

**Pull request merged**